Show automigration message on the Top Site section, fix #2616

gasolin opened the pull request:
UI works, unit test is on the way, I think its ready for the first round review

You can enable this in about:config `browser.migrate.automigrate.inpage.ui.enabled` once `Bug 1361286` is landed

k88hudson commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,51 @@
+        this.store.dispatch(ac.BroadcastToContent({type: at.AUTOMIGRATE_HIDE}));
+        break;
+      case at.AUTOMIGRATE_UNDO_MIGRATION:
+        AutoMigrate.undoAutoMigration(this.window);
```
I think it might be better to use `action._target.browser` (that is, from the `AUTOMIGRATE_UNDO_MIGRATION` action), since there could be multiple windows open at once

gasolin replied:
👍

k88hudson commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,51 @@
+      case at.NEW_TAB_VISIBLE:
+        AutoMigrate.shouldShowMigratePrompt(action._target.browser).then(prompt => {
+          if (prompt) {
+            this.window = action._target.browser.ownerGlobal;
```
See above comment on line 23, this might cause problems since there could be multiple windows open at once

gasolin replied:
✅

k88hudson commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,51 @@
+              }));
+            }
+          } else {
+            this.store.dispatch(ac.BroadcastToContent({type: at.AUTOMIGRATE_HIDE}));
```
Could you perhaps do 
```
else if (this.store.getState().Automigrate.display) { ... }
```
to prevent unnecessary messages/updates?

gasolin replied:
👍

Mardak commented on system-addon/content-src/components/AutoMigratePrompt/AutoMigratePrompt.jsx:
```diff
@@ -0,0 +1,32 @@
+      (<div className="migrate-prompt">
+        <span>{this.props.AutoMigrate.msg}</span>
+        <div className="confirm">
+          <button onClick={this.props.onUndoClick.bind(this)}><FormattedMessage id="migrate_undo_button" /></button>
```
We should probably not do the pure function and bind the function on constructor as binding from `render` will cause render optimizations to fail:
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md

Mardak replied:
Actually, if `onUndoClick` is a prop, `this` probably shouldn't be `AutoMigratePrompt` anyway?

k88hudson replied:
actually yeah, the `this` isn't needed at all

gasolin commented:
Here's how to do the manual testing:

* enable the pref `browser.migrate.automigrate.inpage.ui.enabled` in about:config, 
and `./mach run` with new profile

1. run `./mach run -P` to create a new profile named 'something', then exit without running
2. run `./mach run -P something --migration` to trigger the automigration UI

gasolin commented:
tests added, ready for review again 😊

gasolin commented on system-addon/content-src/components/AutoMigratePrompt/AutoMigratePrompt.jsx:
```diff
@@ -0,0 +1,32 @@
+      </div>
+    </div>) : (<div className="migrate-prompt">
+      <FormattedMessage id="migrate_manual_import_msg" />&nbsp;
+      <FormattedMessage id="migrate_manual_import_link_msg" />&nbsp;
```
this message would be a link to trigger the `manual import` wizard, which will be implemented in bug 1361294
For l10n team friendly I think its better to split strings early before we actually implement that function.

Mardak replied:
I see that AutoMigrate will be passing over at least one of the messages. Any reason why it shouldn't pass over the manual import messages too? I do see that parts of it will be linked, but multiple message parts could be passed over too.

gasolin commented:
@k88hudson or @Mardak could you help review this patch?

Mardak commented on system-addon/content-src/components/AutoMigratePrompt/AutoMigratePrompt.jsx:
```diff
@@ -0,0 +1,32 @@
+
+  return props.AutoMigrate.stage === 0 ?
+    (<div className="migrate-prompt">
+      <span>{props.AutoMigrate.msg}</span>
```
Just making sure, this `msg` should be the success message, e.g., "Dive right into Firefox! Import your favorite sites, bookmarks, history and passwords from Chrome" ? Maybe a more descriptive name will help -- `migratedMessage` ?

Mardak commented on locales/en-US/strings.properties:
```diff
@@ -117,3 +117,11 @@ pocket_feedback_header=The best of the web, curated by over 25 million people.
+# LOCALIZATION NOTE (migrate_manual_import_link_msg): This is shown as the link in message
+migrate_manual_import_link_msg=import your information
+migrate_manual_import_msg_trail=later.
+migrate_import_button=Import It
```
Should this just be "Import" from the spec?

Mardak commented on system-addon/common/Reducers.jsm:
```diff
@@ -91,7 +96,39 @@ function TopSites(prevState = INITIAL_STATE.TopSites, action) {
+    case at.AUTOMIGRATE_IS_REVERTED: {
+      return {
+        display: true,
+        stage: 1,
```
@k88hudson not sure if you have preferences on how this `stage` / state should be handled. With just numbers as here? Or strings or enum-ish? Or I suppose completely different is just deriving both `display` and `stage` from the component and storing the `state` as "HIDDEN" "MIGRATED" "REVERTED" ?

Mardak commented on system-addon/content-src/components/AutoMigratePrompt/AutoMigratePrompt.jsx:
```diff
@@ -0,0 +1,32 @@
+  state => ({AutoMigrate: state.AutoMigrate}),
+  dispatch => ({
+    onImportClick: () => dispatch(ac.SendToMain({type: at.AUTOMIGRATE_MIGRATE_DONE})),
+    onUndoClick: () => dispatch(ac.SendToMain({type: at.AUTOMIGRATE_UNDO_MIGRATION}))
```
I suppose following this pattern, we would want to add a `onManualImportClick` ?

gasolin replied:
yes, it will be in the followup patch

Mardak commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,50 @@
+      case at.NEW_TAB_VISIBLE:
+        AutoMigrate.shouldShowMigratePrompt(action._target.browser).then(prompt => {
+          if (prompt) {
+            let browserName = Services.prefs.getStringPref("browser.migrate.automigrate.browser", "");
```
Why is there this pref logic when home/newtab usage doesn't check? https://dxr.mozilla.org/mozilla-central/search?q=%2Bref%3A%22AutoMigrate%23shouldShowMigratePrompt%22

The logic seems a bit odd in having Activity Stream seemingly randomly check a pref to do something different and somewhat overriding the just called `shouldShowMigratePrompt`. Is it possible to have that `should` method indicate what type of prompt should be shown instead of that a prompt should be shown?

Mardak commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,50 @@
+                data: {msg}
+              }));
+            }
+          } else if (this.store.getState().Automigrate.display) {
```
This looks like a typo? Uncapitalized "M" in `AutoMigrate` ?

Mardak commented on system-addon/lib/AutoMigrateFeed.jsm:
```diff
@@ -0,0 +1,50 @@
+        AutoMigrate.undoAutoMigration(action._target.browser.ownerGlobal);
+        this.store.dispatch(ac.BroadcastToContent({type: at.AUTOMIGRATE_IS_REVERTED}));
+        break;
+      case at.NEW_TAB_VISIBLE:
```
Overall this logic seems a bit strange in that on visibility of a single new tab, there's broadcasts to all tabs?

Mardak commented on locales/en-US/strings.properties:
```diff
@@ -117,3 +117,11 @@ pocket_feedback_header=The best of the web, curated by over 25 million people.
+migrate_manual_import_msg=No problem. We can start you off with a few popular sites instead. You can always
+# LOCALIZATION NOTE (migrate_manual_import_link_msg): This is shown as the link in message
+migrate_manual_import_link_msg=import your information
+migrate_manual_import_msg_trail=later.
```
I see that flod says it won't be difficult to localize, so maybe we just need to fix the implicit concatenation issue with {replacement}. Although actually given that we have a before-link and after-link text, that should provide localizers with enough flexibility to work with? @flodolo

Mardak replied:
Ehh. I could see some locales wanting to just put "." in the trailing text, so if there's an implicit space to join, it would appear as "Stuff stuff later [link] ."

flodolo replied:
Empty spaces in .properties files are a pain. I would go with:

```
# LOCALIZATION NOTE (migrate_manual_import_link_msg): do not translate {link},
# it's replaced by an active link using migrate_manual_import_link as text
migrate_manual_import_msg=No problem. We can start you off with a few popular sites instead. You can always {link} later.
migrate_manual_import_link=import your information
```

gasolin commented:
Rebased and added manual import wizard link PR. @Mardak @k88hudson Due to priority shift I have been asked to help on other photon things. So I will not continue to finish these patch. Still thanks for your help!

k88hudson commented:
Closing this as automigration is no longer being impelmented as such.
